provider/sidecar.go

```diff
@@ -70,150 +89,171 @@ func (provider *Sidecar) Provide(configurationChan chan<- types.ConfigMessage, p
 	}
-	err = provider.loadSidecarConfig(states.ByService())
+
+	err = provider.loadConfig(states.ByService())
 	if err != nil {
-		return err
+		return fmt.Errorf("Error loading Sidecar config: %s", err)
 	}
+
 	return nil
 }
 
-func (provider *Sidecar) constructConfig(sidecarStates map[string][]*service.Service) (*types.Configuration, error) {
-	log.Infoln("loading sidecar config")
-	sidecarConfig := types.Configuration{Backends: provider.makeBackends(sidecarStates)}
-	var err error
-	sidecarConfig.Frontends, err = provider.makeFrontend()
-	if err != nil {
-		return nil, err
+func (provider *Sidecar) loadConfig(sidecarStates map[string][]*service.Service) error {
+	log.Info("Loading sidecar config...")
+
+	config := &types.Configuration{
+		Backends: make(map[string]*types.Backend),
 	}
-	return &sidecarConfig, nil
-}
 
-func (provider *Sidecar) loadSidecarConfig(sidecarStates map[string][]*service.Service) error {
-	conf, err := provider.constructConfig(sidecarStates)
-	if err != nil {
+	if _, err := toml.DecodeFile(provider.Filename, config); err != nil {
 		return err
 	}
+
+	// Create backends from Sidecar state data
+	for serviceName, services := range sidecarStates {
+		backend, ok := config.Backends[serviceName]
+		if !ok {
+			backend = &types.Backend{}
+			config.Backends[serviceName] = backend
+		}
+
+		if backend.LoadBalancer == nil {
+			backend.LoadBalancer = &types.LoadBalancer{Method: defaultMethod, Sticky: defaultSticky}
+		}
+		if backend.Servers == nil {
+			backend.Servers = make(map[string]types.Server)
+		}
+
+		for _, serv := range services {
+			if serv.IsAlive() {
+				ipAddr, err := net.LookupIP(serv.Hostname)
+
+				for i := 0; i < len(serv.Ports); i++ {
+					// TODO: is there any point to add unreachable hosts?
+					var hostname string
+					if err != nil {
+						log.Warn("Failed to resolve IP address: ", err)
+						hostname = serv.Hostname
+					} else {
+						hostname = ipAddr[0].String()
+					}
+
+					backend.Servers[serv.Hostname] = types.Server{
+						URL: fmt.Sprintf("http://%s:%d", hostname, serv.Ports[i].Port),
+					}
+				}
+			}
+		}
+	}
+
 	provider.configurationChan <- types.ConfigMessage{
 		ProviderName:  "sidecar",
-		Configuration: conf,
+		Configuration: config,
 	}
+
+	log.Info("Finished loading sidecar config")
+
 	return nil
 }
 
-func (provider *Sidecar) sidecarWatcher() error {
-	//set timeout to be just a bot more than connection refresh interval
+func (provider *Sidecar) runSidecarWatcher() {
+	// Set timeout to be just a bit more than connection refresh interval
 	provider.connTimer = time.NewTimer(time.Duration(provider.RefreshConn))
-	tr := &http.Transport{ResponseHeaderTimeout: 0}
-	client := &http.Client{
-		Timeout:   0,
-		Transport: tr}
+
 	log.Debugf("Using %s Sidecar connection refresh interval", provider.RefreshConn)
-	provider.recycleConn(client, tr)
-	return nil
+	for {
+		// Call a separate function because the defer statement has function scope
+		provider.sidecarWatcher()
```

Comment:
I have read the original code multiple times now and I still don't feel confident enough to explain to somebody else why it's 100% correct. It may very well be, but I don't understand it fully and I'm probably missing some context. The fact that it is working in production is encouraging so I don't mind leaving it the way you designed it. In this sense, I have created #8 for moving forward with the MaxConn additions.

My objection to the way it was done is clarity and an explanation of why it works the way it does that we can relate to some existing documentation. The fact that `Close()` is never called on that HTTP response body deviates from a well-established Go pattern and the whole pointer reuse technique makes me scratch my head, because the variable will get overwritten during the subsequent iteration of the `for` loop.